### PR TITLE
Add %%COURSE_START_DATE%% to supported keywords

### PR DESCRIPTION
**Pull request by kluo**
Also changes the keyword modal UI to be dynamic

@caesar2164 @stvstnfrd

### CODE REVIEW

File: lms/templates/instructor/instructor_dashboard_2/keyword_sub_modal.html [hunk 1]
```diff
@@ -25,22 +26,12 @@ <h2 id="keyword-sub-tip-title">
...
-      </tr>
-      <tr>
-        <td><input type="text" value="%%COURSE_DISPLAY_NAME%%" readonly="readonly" /></td>
-        <td>${_("display name of the course")}</td>
-      </tr>
-      <tr>
-        <td><input type="text" value="%%COURSE_END_DATE%%" readonly="readonly" /></td>
-        <td>${_("end date of the course")}</td>
-      </tr>
+      % for keyword in keywords_supported:
```
Comment (caesar2164) [inline lms/templates/instructor/instructor_dashboard_2/keyword_sub_modal.html]: YAY for loops!

**Comment by caesar2164** [comment]
Can you edit the commit message to also say that you've converted the keywords to using a loop?
(ideally you'd actually split this into two commits, one for the looping and other background things, and one for adding the new keyword, but I don't think it's worth doing that now...)

**Comment by caesar2164** [comment]
how about:

```Make Keyword Modal more dynamic & add new keyword

- make keyword sub modal loop through dict of available keywords
- change position of keyword modal in HTML
- add %%COURSE_START_DATE%% keyword
```

File: cms/static/js/views/utils/view_utils.js [hunk 2]
```diff
@@ -175,7 +175,7 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
 
         var keywordValidator = (function () {
             var regexp = /%%[^%\s]+%%/g;
-            var keywordsSupported = ['%%USER_ID%%', '%%USER_FULLNAME%%', '%%COURSE_DISPLAY_NAME%%', '%%COURSE_END_DATE%%'];
+            var keywordsSupported = ['%%USER_ID%%', '%%USER_FULLNAME%%', '%%COURSE_DISPLAY_NAME%%', '%%COURSE_START_DATE%%', '%%COURSE_END_DATE%%'];
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Given that list line continues to grow, can we split it up and put each item on a separate line?

File: common/djangoapps/util/keyword_substitution.py [hunk 3]
```diff
@@ -57,9 +58,9 @@ def substitute_keywords(string, user=None, course=None):
         # Cannot proceed without course and user information
         return string
 
-    for key, func in KEYWORD_FUNCTION_MAP.iteritems():
+    for key, value in KEYWORD_FUNCTION_MAP.iteritems():
         if key in string:
-            substitutor = func(user, course)
+            substitutor = value[0](user, course)
```
Comment (stvstnfrd) [inline common/djangoapps/util/keyword_substitution.py]: Could this be a named tuple or maybe even a dict? That way we can do something like:

```python
substitutor = value.substitute(user, course)
# or
substitutor = value['substitute'](user, course)
# the use of 'substitute above is an example; you could choose a more apt name
```

I get a bit uneasy about directly indexing into tuples like `value[0]`.

File: lms/djangoapps/instructor/views/instructor_dashboard.py [hunk 4]
```diff
@@ -355,6 +356,7 @@ def _section_send_email(course, access):
     section_data = {
         'section_key': 'send_email',
         'section_display_name': _('Email'),
+        'keywords_supported': tuple((keyword, value[1]) for keyword, value in keyword_substitution.KEYWORD_FUNCTION_MAP.iteritems()),
```
Comment (stvstnfrd) [inline lms/djangoapps/instructor/views/instructor_dashboard.py]: See note above about named tuples.

File: lms/startup.py [hunk 5]
```diff
@@ -190,10 +194,11 @@ def course_end_date_sub(user, course):
...
     kf_map = {
-        '%%USER_ID%%': user_id_sub,
-        '%%USER_FULLNAME%%': user_fullname_sub,
-        '%%COURSE_DISPLAY_NAME%%': course_display_name_sub,
-        '%%COURSE_END_DATE%%': course_end_date_sub
+        '%%USER_ID%%': (user_id_sub, 'anonymous_user_id (for use in survey links)'),
+        '%%USER_FULLNAME%%': (user_fullname_sub, 'user profile name'),
+        '%%COURSE_DISPLAY_NAME%%': (course_display_name_sub, 'display name of the course'),
+        '%%COURSE_START_DATE%%': (course_start_date_sub, 'start date of the course'),
+        '%%COURSE_END_DATE%%': (course_end_date_sub, 'end date of the course')
```
Comment (stvstnfrd) [inline lms/startup.py]: Add a trailing comma.

File: lms/static/coffee/src/instructor_dashboard/util.coffee [hunk 6]
```diff
@@ -303,7 +303,7 @@ class PendingInstructorTasks
 class KeywordValidator
 
     @keyword_regex = /%%+[^%]+%%/g
-    @keywords = ['%%USER_ID%%', '%%USER_FULLNAME%%', '%%COURSE_DISPLAY_NAME%%', '%%COURSE_END_DATE%%']
+    @keywords = ['%%USER_ID%%', '%%USER_FULLNAME%%', '%%COURSE_DISPLAY_NAME%%', '%%COURSE_START_DATE%%', '%%COURSE_END_DATE%%']
```
Comment (stvstnfrd) [inline lms/static/coffee/src/instructor_dashboard/util.coffee]: See note above about splitting onto multiple lines.

File: lms/templates/instructor/instructor_dashboard_2/keyword_sub_modal.html [hunk 7]
```diff
@@ -1,4 +1,5 @@
 <%! from django.utils.translation import ugettext as _ %>
+<%page args="keywords_supported"/>
```
Comment (stvstnfrd) [inline lms/templates/instructor/instructor_dashboard_2/keyword_sub_modal.html]: Should there be a space after `%`?

File: lms/templates/instructor/instructor_dashboard_2/keyword_sub_modal.html [hunk 8]
```diff
@@ -25,22 +26,12 @@ <h2 id="keyword-sub-tip-title">
...
-        <td>${_("display name of the course")}</td>
-      </tr>
-      <tr>
-        <td><input type="text" value="%%COURSE_END_DATE%%" readonly="readonly" /></td>
-        <td>${_("end date of the course")}</td>
-      </tr>
+      % for keyword in keywords_supported:
+        <tr>
+          <td><input type="text" value="${keyword[0]}" readonly="readonly" /></td>
+          <td>${_(keyword[1])}</td>
```
Comment (stvstnfrd) [inline lms/templates/instructor/instructor_dashboard_2/keyword_sub_modal.html]: See note above about named tuples.

File: lms/templates/instructor/instructor_dashboard_2/send_email.html [hunk 9]
```diff
@@ -1,5 +1,6 @@
 <%! from django.utils.translation import ugettext as _ %>
 <%page args="section_data"/>
+<%include file='keyword_sub_modal.html' args="keywords_supported=section_data['keywords_supported']" />
```
Comment (stvstnfrd) [inline lms/templates/instructor/instructor_dashboard_2/send_email.html]: Should there be a space after `%`?

File: lms/templates/instructor/instructor_dashboard_2/instructor_dashboard_2.html [hunk 10]
```diff
@@ -116,6 +116,3 @@
     </section>
   </div>
 </section>
-
-## Modal included here so leanModal can position it outside of dashboard wrapper
-<%include file='keyword_sub_modal.html' />
```
Comment (stvstnfrd) [inline lms/templates/instructor/instructor_dashboard_2/instructor_dashboard_2.html]: Why are we moving this?

Comment (kluo) [inline lms/templates/instructor/instructor_dashboard_2/instructor_dashboard_2.html]: There's not a clean way to pass it the context variable for keywords_supported here

File: lms/templates/instructor/instructor_dashboard_2/send_email.html (hunk 9 above)
Comment (kluo) [inline lms/templates/instructor/instructor_dashboard_2/send_email.html]: No, this the mako syntax

File: lms/templates/instructor/instructor_dashboard_2/keyword_sub_modal.html (hunk 7 above)
Comment (kluo) [inline lms/templates/instructor/instructor_dashboard_2/keyword_sub_modal.html]: This is the Mako syntax

File: common/djangoapps/util/keyword_substitution.py (hunk 3 above)
Comment (kluo) [inline common/djangoapps/util/keyword_substitution.py]: Didn't know about namedtuples, I think I'll go for that one

**Comment by stvstnfrd** [comment]
+1

**Pull request merged**